200/step2_2_bfs.py

```diff
@@ -0,0 +1,35 @@
+from typing import List, Tuple
+from collections import deque
+from itertools import product
+
+
+class Solution:
+    def numIslands(self, grid: List[List[str]]) -> int:
+        # gridを破壊的に変更するので注意！！
+        LAND_CELL = '1'
+        WATER_CELL = '0'
+        if not grid or not grid[0]:
+            return 0
+
+        rows, cols = len(grid), len(grid[0])
+
+        def is_in_bounds(row: int, column: int) -> bool:
+            return 0 <= row < rows and 0 <= column < cols
+
+        def sink_island_with_bfs(start_row: int, start_column: int) -> None:
+            cells_to_visit: deque[Tuple[int, int]] = deque([(start_row, start_column)])
+            grid[start_row][start_column] = WATER_CELL
+            while cells_to_visit:
+                row, column = cells_to_visit.popleft()
+                for direction_row, direction_column in ((-1, 0), (1, 0), (0, -1), (0, 1)):
```

Comment:
((-1, 0), (1, 0), (0, -1), (0, 1))ですが、同類の繰り返しはタプルよりリストの方が自然という意見をどこかで見た気がします。

```suggestion
                directions = [
                    (-1, 0),
                    (1, 0),
                    (0, -1),
                    (0, 1)
                ]
                for direction_row, direction_column in directions:
```
とかでもいいかもしれません。（趣味の範囲だと思います。）

また、directionsは定数なので、関数内の最上部で定義してもいいかもしれません。

Reply:
docto-rinさん、ご指摘ありがとうございます
おっしゃるように、提案いただいた書き方のほうが行の長さも短くなり、認知負荷が下がりそうです。